src/achievements.cpp

```diff
@@ -0,0 +1,107 @@
+/*
+	Copyright (C) 2003 - 2022
+	by David White <dave@whitevine.net>
+	Part of the Battle for Wesnoth Project https://www.wesnoth.org/
+
+	This program is free software; you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation; either version 2 of the License, or
+	(at your option) any later version.
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY.
+
+	See the COPYING file for more details.
+*/
+
+#include "achievements.hpp"
+
+#include "filesystem.hpp"
+#include "game_config.hpp"
+#include "log.hpp"
+#include "preferences/general.hpp"
+#include "serialization/parser.hpp"
+#include "serialization/preprocessor.hpp"
+
+static lg::log_domain log_config("config");
+#define ERR_CONFIG LOG_STREAM(err, log_config)
+
+// TODO: some way to display in-game when something is achieved?
+
+// TODO: add to schema
+// TODO: unit tests?
+
+/**
+ * Reads the mainline achievements.cfg and then all the achievements of each installed add-on.
+ *
+ * This is intentionally handled separately from other WML loading so that:
+ * a) All achievements and their status are able to be displayed on the main menu right after Wesnoth starts and regardless of which add-ons are active.
```

Comment:
This seems overly complex, given that all add-on's `[campaign]`, `[modification]` or similar tags already have to be available at the title screen. I think this should be able to fit in with that mechanism.

Reply:
That's not true at all. Only add-ons intended for use in single-player need to be visible at the title screen. Anything written as multiplayer-only does not have to be (and usually isn't).

Reply:
Ah true, anything inside `#ifdef MULTIPLAYER` will be missing. However, instead of having to have an extra WML-loading routine in C++, would it be better for any multiplayer add-ons that include achievements to have some achievement-specific WML outside their `#ifdef MULTIPLAYER`block?

Reply:
That would still limit the viewable achievements when playing a game, since add-ons that aren't used are inactive, wouldn't it?

Reply:
True. I'll mark this resolved.